provider/src/main/kotlin/org/gradle/kotlin/dsl/precompile/PrecompiledProjectScript.kt

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.kotlin.dsl.precompile
+
+import org.gradle.api.Project
+
+import org.gradle.kotlin.dsl.GradleDsl
+import org.gradle.kotlin.dsl.ScriptHandlerScope
+
+import org.gradle.plugin.use.PluginDependenciesSpec
+import org.gradle.plugin.use.PluginDependencySpec
+
+import kotlin.script.extensions.SamWithReceiverAnnotations
+import kotlin.script.templates.ScriptTemplateDefinition
+
+
+/**
+ * Script template definition for precompiled Kotlin script with a [Project] target.
+ */
+@ScriptTemplateDefinition(
+    resolver = PrecompiledScriptDependenciesResolver::class,
+    scriptFilePattern = "^.*\\.gradle\\.kts$")
+@SamWithReceiverAnnotations("org.gradle.api.HasImplicitReceiver")
+@GradleDsl
+abstract class PrecompiledProjectScript(project: Project) : Project by project {
```

Comment:
A comment about why this is needed and why it's different than the `ProjectScript` would be really useful.
Eg. 
```
These scripts are applied to project scripts.
```

Also why are they called "Precompiled"?

Reply:
They are called "Precompiled" because those script templates are for Kotlin scripts compiled ahead of time as part of regular Kotlin sourcesets and then distributed as regular binary plugins. Does the name make more sense now?

Reply:
Yes, documenting that would be useful.